File: platform/Emscripten/src/binary_model_wire_codec.js
const RECORD_HEADER_BYTES = 12;
const FINGERPRINT_MULTIPLIER = 1315423911n;
const STRING_FINGERPRINT_MULTIPLIER = 1099511628211n;

const scratchBuffer = new ArrayBuffer(4);
const scratchView = new DataView(scratchBuffer);
const textDecoder = new TextDecoder();

function writeU32LE(bytes, offset, value) {
  bytes[offset] = value & 0xff;
  bytes[offset + 1] = (value >>> 8) & 0xff;
  bytes[offset + 2] = (value >>> 16) & 0xff;
  bytes[offset + 3] = (value >>> 24) & 0xff;
}

function readU32LE(dataView, offset) {
  return dataView.getUint32(offset, true);
}

function floatBits(value) {
  scratchView.setFloat32(0, Math.fround(value), true);
  return scratchView.getUint32(0, true);
}

function floatFromBits(bits) {
  scratchView.setUint32(0, bits, true);
  return scratchView.getFloat32(0, true);
}

function buildAsciiString(index, targetBytes) {
  let value = `model_${index}_`;
  if (value.length > targetBytes) {
    return value.slice(0, targetBytes);
  }

  while (value.length < targetBytes) {
    value += String.fromCharCode(97 + ((index + value.length) % 26));
  }
  return value;
}

function makeBinaryModel(index, stringBytes) {
  const floatValue = Math.fround(0.5 + index * 0.25);
  return {
    intValue: 1000 + index * 17,
    floatValue,
    floatBits: floatBits(floatValue),
    stringValue: buildAsciiString(index, stringBytes),
  };
}

function createModelsForTargetBytes(targetBytes) {
  if (targetBytes < RECORD_HEADER_BYTES) {
    throw new Error("Target payload is too small for BinaryModel.");
  }

  const models = [];
  let usedBytes = 0;
  let index = 0;

  while (usedBytes < targetBytes) {
    const remaining = targetBytes - usedBytes;
    if (remaining < RECORD_HEADER_BYTES) {
      if (models.length === 0) {
        throw new Error("BinaryModel payload generation failed.");
      }
      models[models.length - 1].stringValue += buildAsciiString(index, remaining);
      usedBytes += remaining;
      break;
    }

    const maxStringBytes = remaining - RECORD_HEADER_BYTES;
    let stringBytes = Math.min(maxStringBytes, 12 + (index % 19));
    const tailBytes = remaining - RECORD_HEADER_BYTES - stringBytes;
    if (tailBytes > 0 && tailBytes < RECORD_HEADER_BYTES) {
      stringBytes += tailBytes;
    }

    models.push(makeBinaryModel(index, stringBytes));
    usedBytes += RECORD_HEADER_BYTES + stringBytes;
    index += 1;
  }

  if (encodedSize(models) !== targetBytes) {
    throw new Error("BinaryModel payload generation did not hit the requested size.");
  }

  return models;
}

function encodedSize(models) {
  let totalBytes = 0;
  for (const model of models) {
    totalBytes += RECORD_HEADER_BYTES + model.stringValue.length;
  }
  return totalBytes;
}

function encodeTo(models, bytes, offset = 0) {
  let cursor = offset;
  for (const model of models) {
    writeU32LE(bytes, cursor, model.intValue >>> 0);
    cursor += 4;
    writeU32LE(bytes, cursor, model.floatBits >>> 0);
    cursor += 4;
    writeU32LE(bytes, cursor, model.stringValue.length);
    cursor += 4;

    for (let index = 0; index < model.stringValue.length; index += 1) {
      bytes[cursor + index] = model.stringValue.charCodeAt(index) & 0xff;
    }
    cursor += model.stringValue.length;
  }
}

function encode(models) {
  const bytes = new Uint8Array(encodedSize(models));
  encodeTo(models, bytes);
  return bytes;
}

function encodeToHeap(models, module, pointer) {
  const bytes = module.HEAPU8;
  encodeTo(models, bytes, pointer);
}

function decode(bytes) {
  const dataView = new DataView(bytes.buffer, bytes.byteOffset, bytes.byteLength);
  const models = [];
  let offset = 0;

  while (offset < bytes.byteLength) {
    const intValue = readU32LE(dataView, offset) | 0;
    offset += 4;
    const bits = readU32LE(dataView, offset);
    offset += 4;
    const stringSize = readU32LE(dataView, offset);
    offset += 4;

    if (offset + stringSize > bytes.byteLength) {
      throw new Error("BinaryModel string exceeds compact payload size.");
    }

    const stringValue = textDecoder.decode(bytes.subarray(offset, offset + stringSize));
    offset += stringSize;

    models.push({
      intValue,
      floatValue: floatFromBits(bits),
      floatBits: bits,
      stringValue,
    });
  }

  return models;
}

function fingerprint(models) {
  if (models.length === 0) {
    return 0n;
  }

  const middle = Math.floor(models.length / 2);
  const samples = [models[0], models[middle], models[models.length - 1]];
  let value = BigInt(models.length);

  for (const model of samples) {
    value = BigInt.asUintN(64, (value * FINGERPRINT_MULTIPLIER) ^ BigInt(model.intValue >>> 0));
    value = BigInt.asUintN(64, (value * FINGERPRINT_MULTIPLIER) ^ BigInt(model.floatBits >>> 0));
    value = BigInt.asUintN(64, (value * FINGERPRINT_MULTIPLIER) ^ BigInt(model.stringValue.length));
    for (let index = 0; index < model.stringValue.length; index += 1) {
      value = BigInt.asUintN(
        64,
        (value * STRING_FINGERPRINT_MULTIPLIER) ^ BigInt(model.stringValue.charCodeAt(index) & 0xff),
      );
    }
  }

  return value;
}

module.exports = {
  createModelsForTargetBytes,
  decode,
  encode,
  encodeToHeap,
  encodedSize,
  fingerprint,
};

File: platform/Android/app/src/main/java/com/ffibb/android/BinaryModelWireCodec.java
package com.ffibb.android;

import java.nio.ByteBuffer;
import java.nio.ByteOrder;
import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.List;

final class BinaryModelWireCodec {
  static final int RECORD_HEADER_BYTES = 4 + 4 + 4;

  private BinaryModelWireCodec() {
  }

  static byte[] encode(List<BinaryModel> models) {
    byte[] bytes = new byte[encodedSize(models)];
    int offset = 0;

    for (BinaryModel model : models) {
      writeIntLE(bytes, offset, model.intValue);
      offset += 4;
      writeIntLE(bytes, offset, Float.floatToRawIntBits(model.floatValue));
      offset += 4;

      byte[] stringBytes = model.stringValue.getBytes(StandardCharsets.UTF_8);
      writeIntLE(bytes, offset, stringBytes.length);
      offset += 4;

      System.arraycopy(stringBytes, 0, bytes, offset, stringBytes.length);
      offset += stringBytes.length;
    }

    return bytes;
  }

  static void encodeToBuffer(List<BinaryModel> models, ByteBuffer output) {
    int size = encodedSize(models);
    if (output.capacity() < size) {
      throw new IllegalArgumentException("Direct buffer is too small for compact payload.");
    }

    output.clear();
    output.order(ByteOrder.LITTLE_ENDIAN);

    for (BinaryModel model : models) {
      output.putInt(model.intValue);
      output.putInt(Float.floatToRawIntBits(model.floatValue));

      byte[] stringBytes = model.stringValue.getBytes(StandardCharsets.UTF_8);
      output.putInt(stringBytes.length);
      output.put(stringBytes);
    }

    output.flip();
  }

  static List<BinaryModel> decode(byte[] bytes) {
    ArrayList<BinaryModel> models = new ArrayList<>();
    int offset = 0;

    while (offset < bytes.length) {
      if (bytes.length - offset < RECORD_HEADER_BYTES) {
        throw new IllegalArgumentException("Unexpected end of compact byte payload.");
      }

      int intValue = readIntLE(bytes, offset);
      offset += 4;

      float floatValue = Float.intBitsToFloat(readIntLE(bytes, offset));
      offset += 4;

      int stringSize = readIntLE(bytes, offset);
      offset += 4;

      if (stringSize < 0 || bytes.length - offset < stringSize) {
        throw new IllegalArgumentException("String size exceeds compact payload size.");
      }

      String stringValue = new String(bytes, offset, stringSize, StandardCharsets.UTF_8);
      offset += stringSize;
      models.add(new BinaryModel(intValue, floatValue, stringValue));
    }

    return models;
  }

  static List<BinaryModel> decode(ByteBuffer bytes) {
    ByteBuffer input = bytes.duplicate();
    input.order(ByteOrder.LITTLE_ENDIAN);

    ArrayList<BinaryModel> models = new ArrayList<>();
    while (input.hasRemaining()) {
      if (input.remaining() < RECORD_HEADER_BYTES) {
        throw new IllegalArgumentException("Unexpected end of compact byte payload.");
      }

      int intValue = input.getInt();
      float floatValue = Float.intBitsToFloat(input.getInt());
      int stringSize = input.getInt();

      if (stringSize < 0 || input.remaining() < stringSize) {
        throw new IllegalArgumentException("String size exceeds compact payload size.");
      }

      byte[] stringBytes = new byte[stringSize];
      input.get(stringBytes);
      models.add(new BinaryModel(intValue, floatValue, new String(stringBytes, StandardCharsets.UTF_8)));
    }

    return models;
  }

  static int encodedSize(List<BinaryModel> models) {
    int totalBytes = 0;
    for (BinaryModel model : models) {
      totalBytes += RECORD_HEADER_BYTES + model.stringValue.getBytes(StandardCharsets.UTF_8).length;
    }
    return totalBytes;
  }

  static long fingerprint(List<BinaryModel> models) {
    if (models.isEmpty()) {
      return 0L;
    }

    int middle = models.size() / 2;
    BinaryModel[] samples = new BinaryModel[] {
        models.get(0),
        models.get(middle),
        models.get(models.size() - 1)
    };

    long fingerprint = models.size();
    for (BinaryModel model : samples) {
      fingerprint = (fingerprint * 1315423911L) ^ (model.intValue & 0xFFFFFFFFL);
      fingerprint = (fingerprint * 1315423911L) ^ (Float.floatToRawIntBits(model.floatValue) & 0xFFFFFFFFL);

      byte[] stringBytes = model.stringValue.getBytes(StandardCharsets.UTF_8);
      fingerprint = (fingerprint * 1315423911L) ^ stringBytes.length;
      for (byte value : stringBytes) {
        fingerprint = (fingerprint * 1099511628211L) ^ (value & 0xFFL);
      }
    }

    return fingerprint;
  }

  private static void writeIntLE(byte[] bytes, int offset, int value) {
    bytes[offset] = (byte) (value & 0xFF);
    bytes[offset + 1] = (byte) ((value >>> 8) & 0xFF);
    bytes[offset + 2] = (byte) ((value >>> 16) & 0xFF);
    bytes[offset + 3] = (byte) ((value >>> 24) & 0xFF);
  }

  private static int readIntLE(byte[] bytes, int offset) {
    return (bytes[offset] & 0xFF) |
        ((bytes[offset + 1] & 0xFF) << 8) |
        ((bytes[offset + 2] & 0xFF) << 16) |
        ((bytes[offset + 3] & 0xFF) << 24);
  }
}
